Added `skip_absent()` feature to `setnames()`

MusTheDataGuy opened the pull request:
Closes #3030 

Added `skip_absent()` feature to `setnames()` so that elements not present in `old` won't cause the function to exit but will instead skip onto the next element.

**`skip_absent()` is set to `FALSE` by default and thus will not affect the default behaviour of `setnames()`.**

When building this element, I became aware of some issues when building the package on a Windows 10 machine that related to `.dll` file conflicts, and as such updated `.gitignore` and `.Rbuildignore` which rectified the problem:

* Added `*.dll` to `.gitignore`
* Added `^.*\.dll$` to `.Rbuildignore`

I initially intended to add the `isTrue(skip_absent)` check as a standalone element within `setnames()` but found that it kept failing due to `i` being `NA`, which subsequently halted the `setnames()` function.

From there it was clear that the solution was to incorporate the `isTrue(skip_absent)` check *within* the `if(anyNA(i)) { ... }` check that occurs so that it is called when `i` is `NA` (i.e. absent).  This in turn indicates that said element is not present, which logically implies that this is the correct place to check for any missing elements of `old` and to skip accordingly.

I would also like to thank @mattdowle for his assistance throughout this process.

MichaelChirico commented on NEWS.md:
```diff
@@ -1,13 +1,14 @@
 #### NEW FEATURES
 
-1. `fread()` can now read a remote compressed file in one step; `fread("https://domain.org/file.csv.bz2")`. The `file=` argument now supports `.gz` and `.bz2` too; i.e. `fread(file="file.csv.gz")` works now where only `fread("file.csv.gz")` worked in 1.11.8.
+1. In those cases where you need to rename columns in a `DT` but the columns aren't always known, `setnames()` now contains an additional argument (`skip_absent`) to skip them if they aren't present.  For example, if you know that columns `a`, `b` and `d` are present in `DT`, but you don't know if column `c` is or isn't, then you can include `c` in `old` and if it isn't found, `setnames()` will s ...
```
Please add new `NEWS` items at the bottom

MichaelChirico commented on R/data.table.R:
```diff
@@ -2532,7 +2531,7 @@ setnames <- function(x,old,new) {
       if (any(duplicated(old))) stop("Some duplicates exist in 'old': ", paste(old[duplicated(old)],collapse=","))
       i = chmatch(old,names(x))
-      if (anyNA(i)) stop("Items of 'old' not found in column names: ",paste(old[is.na(i)],collapse=","))
+      if (anyNA(i)){ if (skip_absent == TRUE){ w <- old %chin% names(x); old = old[w]; new = new[w]; i = i[w] } else { stop("Items of 'old' not found in column names: ",paste(old[is.na(i)],collapse=",")) } }
```
please split the branch logic onto new lines, it's a bit hard to read at the moment

MichaelChirico commented on inst/tests/tests.Rraw:
```diff
@@ -12328,6 +12328,11 @@ test(1951.4, d1[d2, nomatch=3], error="nomatch= must be either NA or NULL .or 0
+DT <- data.table(a = 1, b = 2, d = 3)
+old <- c("a", "b", "c", "d")
+new <- c("A", "B", "C", "D") 
+test(1953, setnames(DT, old, new, skip_absent = TRUE), DT <- data.table(A = 1, B = 2, D = 3))
```
please also test potential erroneous behavior/mis-use of `skip_absent`

mattdowle replied:
agree need more tests please.  `skip_absent` should only accept `TRUE` and `FALSE` and fail with error on anything else (use `stopifnot`).  There should be a test when TRUE and all are missing, when TRUE and none are missing, when TRUE and two are missing, and when there are missings and the non-missings in old don't correspond to names(DT) in the same order.   The `DT <-` in the y part of test 1953 can be removed too.

MichaelChirico commented:
Thanks for the PR!

MichaelChirico commented:
I'm seeing a big whitespace diff on `.gitignore`... you're on Windows right? I suspect your editor is inserting `\r\n` newlines instead of the standard `\n`... could you try and confirm and, if so, stick to `\n`? 😅

MichaelChirico commented:
Could you remove the new `NEWS.html`?

MichaelChirico commented on R/data.table.R:
```diff
@@ -2532,7 +2531,13 @@ setnames <- function(x,old,new) {
       if (any(duplicated(old))) stop("Some duplicates exist in 'old': ", paste(old[duplicated(old)],collapse=","))
       i = chmatch(old,names(x))
-      if (anyNA(i)) stop("Items of 'old' not found in column names: ",paste(old[is.na(i)],collapse=","))
+      if (anyNA(i)){ if (skip_absent == TRUE){
```
either `if (skip_absent)` or (generally more preferred) `if (isTRUE(skip_absent))`

MichaelChirico commented on R/data.table.R:
```diff
@@ -2532,7 +2531,14 @@ setnames <- function(x,old,new) {
       i = chmatch(old,names(x))
-      if (anyNA(i)) stop("Items of 'old' not found in column names: ",paste(old[is.na(i)],collapse=","))
+      if (anyNA(i)){ if (skip_absent == TRUE){
+        w <- old %chin% names(x)
```
what happens when `length(w) == 0`?

mattdowle replied:
test 1955.3 now covers that case (all missing)

mattdowle commented on man/setattr.Rd:
```diff
@@ -7,15 +7,17 @@
   \item{value}{ The value to assign to the attribute or \code{NULL} removes the attribute, if present. }
   \item{old}{ When \code{new} is provided, character names or numeric positions of column names to change. When \code{new} is not provided, the new column names, which must be the same length as the number of columns. See examples. }
   \item{new}{ Optional. New column names, must be the same length as columns provided to \code{old} argument. }
+  \item{skip_absent}{ Skips cases where there is no match in \code{old}.  Set to \code{FALSE} by default. Switch flag to \code{TRUE} to activate. }
```
```suggestion
  \item{skip_absent}{ Skip items in \code{old} that are missing (i.e. absent) in `names(x)`. Default \code{FALSE} halts with error if any are missing. }
```

mattdowle commented on man/setattr.Rd:
```diff
@@ -34,6 +36,13 @@ setnames(x,old,new)
+old <- c("a", "b", "c", "d")
+new <- c("A", "B", "C", "D")
+
+setnames(DT, old, new, skip_absent = TRUE) # check for column names in old and skip if item is absent
```
```suggestion
setnames(DT, old, new, skip_absent = TRUE) # skips `old[3]` because `"c"` is not a column name of `DT`
```

mattdowle commented on DESCRIPTION:
```diff
@@ -12,7 +12,8 @@ Authors@R: c(
   person("Hugh","Parsonage",  role="ctb"),
-  person("Scott","Ritchie",   role="ctb"))
+  person("Scott","Ritchie",   role="ctb"),
+  person("Mus","Yaramaz-David", role="ctb"))
```
People don't normally add themselves here. There are 53 contributors who have done something, no matter how small:  https://github.com/Rdatatable/data.table/graphs/contributors. You would be listed there, and in NEWS. When a contributor has made sustained contributions, or a few big ones, I add them to DESCRIPTION which is displayed on CRAN. The contributors listed in DESCRIPTION are roughly listed at the top of https://github.com/Rdatatable/data.table/graphs/contributors.
At least, this is the way this has been done to date.  We could change so that every single contributor is listed in DESCRIPTION if people want to.

MusTheDataGuy replied:
I think that something mid-way between the two would be ideal.

For example, those who make small, non-functional changes (such as correcting typos) wouldn't be included here but those who extend the functionality of `data.table` in some meaningful and/or useful way should be included.

That way it would prevent this file from becoming too large, whilst still retaining its relevance and meaning.

HughParsonage replied:
I think the author field should be limited to those who have contributed code that the authors would not have ordinarily come up with. This PR is an enhancement but its implementation is quite elementary so should not qualify.

mattdowle commented on NEWS.md:
```diff
@@ -7,7 +6,11 @@
+
+4. In those cases where you need to rename columns in a `DT` but the columns aren't always known, `setnames()` now contains an additional argument (`skip_absent`) to skip them if they aren't present.  For example, if you know that columns `a`, `b` and `d` are present in `DT`, but you don't know if column `c` is or isn't, then you can include `c` in `old` and if it isn't found, `setnames()` will s ...
+
+3. In those cases where you need to rename columns in a `DT` but the columns aren't always known, `setnames()` now contains an additional argument (`skip_absent`) to skip them if they aren't present.  For example, if you know that columns `a`, `b` and `d` are present in `DT`, but you don't know if column `c` is or isn't, then you can include `c` in `old` and if it isn't found, `setnames()` will s ...
```
Looks like a duplicate new item here.   Please also add "Thanks to @MusTheDataGuy for the suggestion (#3030) and the PR (#3111)."  (i.e. item 2 in Contributing guide mentions this.)

mattdowle commented:
Looks great. Just a few minor tweaks and then it can be merged (see comments inline). Thanks!

MusTheDataGuy commented:
> I'm seeing a big whitespace diff on `.gitignore`... you're on Windows right? I suspect your editor is inserting `\r\n` newlines instead of the standard `\n`... could you try and confirm and, if so, stick to `\n`? 😅
> 
> ![screen shot 2018-10-17 at 8 55 20 pm](https://user-images.githubusercontent.com/7606389/47087672-2af8d500-d24f-11e8-8939-b70594b01310.png)

That's right, I'm running Windows 10. I have updated the file now and will do this for any other affected files also.

For those who are interested, the way I fixed the problem was to change the setting of my code editor (VS Code).  It is set to `CRLF` by default; I have switched it to `LF` and the problem is now fixed.

HughParsonage commented on R/data.table.R:
```diff
@@ -18,7 +17,7 @@ setPackageName("data.table",.global)
 .SD = .N = .I = .GRP = .BY = .EACHI = NULL
 # These are exported to prevent NOTEs from R CMD check, and checkUsage via compiler.
-# But also exporting them makes it clear (to users and other packages) that data.table uses these as symbols.
+# But also exporting them makes it clear (to users and other packages) that data.table uses these as symbols.`
```
Please delete.

Atrebas commented:
Minor comment, in `?setnames`, the comment on line
```
setnames(DT,"b","B")               # by name, no match() needed (warning if "b" is missing)
```
is not longer true with the `skip_absent` argument.
